### PR TITLE
Add aarch64 and RISCV64 support

### PR DESCRIPTION
**Pull request by rockrush**
Depend on sysfs instead of /proc/cpuinfo, for content of this file varies
across CPU architectures, and doesn't offer unified info tag.

Solves issue #199 .

Signed-off-by: Xingyou Chen <rockrush@rockwork.org>

### CODE REVIEW

**Comment by fromanirh** [comment]
Thanks for this PR!
I'm generally in favour of moving from `/proc/cpuinfo` to `sysfs` to learn about the CPU data.
Let's check this change is covered by the existing tests, and let's add more of them if needed
other than that, at glance looks good! I'll have a deeper look later

File: cmd/ghwc/commands/cpu.go [hunk 1]
```diff
@@ -43,12 +43,12 @@ func showCPU(cmd *cobra.Command, args []string) error {
 				// pretty-print the (large) block of capability strings into rows
 				// of 6 capability strings
 				rows := int(math.Ceil(float64(len(proc.Capabilities)) / float64(6)))
-				for row := 1; row < rows; row = row + 1 {
-					rowStart := (row * 6) - 1
+				for row := 0; row < rows; row = row + 1 {
```
Comment (fromanirh) [inline cmd/ghwc/commands/cpu.go:46]: this change seems to have some merit, but it should be moved on its own commit (or possibly on its own PR) since it seems more refactoring/cleanup than actually part of the core concept of this PR

File: pkg/cpu/cpu_linux.go [hunk 2]
```diff
@@ -64,85 +84,69 @@ func processorsGet(ctx *context.Context) []*Processor {
...
 	}
 
-	// Build a set of physical processor IDs which represent the physical
-	// package of the CPU
-	setPhysicalIDs := make(map[int]bool)
-	for _, attrs := range procAttrs {
-		pid, err := strconv.Atoi(attrs["physical id"])
-		if err != nil {
+	// Iterate on /sys/devices/system/cpu/cpuN, not on /proc/cpuinfo
+	Entries, _ := ioutil.ReadDir(paths.SysDevicesSystemCPU)
```
Comment (fromanirh) [inline pkg/cpu/cpu_linux.go:88]: even if (very) unlikely, should we probably handle the error here

File: pkg/cpu/cpu_linux.go [hunk 3]
```diff
@@ -64,85 +84,69 @@ func processorsGet(ctx *context.Context) []*Processor {
...
-		p.Model = first["model name"]
-		p.Vendor = first["vendor_id"]
-		numCores, err := strconv.Atoi(first["cpu cores"])
-		if err != nil {
-			continue
-		}
-		p.NumCores = uint32(numCores)
-		numThreads, err := strconv.Atoi(first["siblings"])
-		if err != nil {
+		lcoreID, error := strconv.Atoi(lcore.Name()[3:]) // get "N" out of "cpuN"
```
Comment (fromanirh) [inline pkg/cpu/cpu_linux.go:95]: (very) minor: since we are already using a regexp, we can evaluate if we can use a group to extract the cpuid right away instead of doing in two stages. Not blocking, can be done in a later PR/commit

File: pkg/cpu/cpu_linux.go [hunk 4]
```diff
@@ -64,85 +84,69 @@ func processorsGet(ctx *context.Context) []*Processor {
...
-				continue
-			}
-			coreID, err := strconv.Atoi(procAttrs[lpidx]["core id"])
-			if err != nil {
-				continue
+		// Fetch CPU ID
+		physIdPath := fmt.Sprintf(
+			"%s/cpu%d/topology/physical_package_id",
+			paths.SysDevicesSystemCPU,
+			lcoreID,
```
Comment (fromanirh) [inline pkg/cpu/cpu_linux.go:104]: even if this is only relevant for linux, I prefer to use `filepath.Join` when building paths.

File: pkg/cpu/cpu_linux.go [hunk 5]
```diff
@@ -64,85 +84,69 @@ func processorsGet(ctx *context.Context) []*Processor {
...
-			lpid, err := strconv.Atoi(procAttrs[lpidx]["processor"])
-			if err != nil {
-				continue
-			}
-			coreID, err := strconv.Atoi(procAttrs[lpidx]["core id"])
-			if err != nil {
-				continue
+		// Fetch CPU ID
+		physIdPath := fmt.Sprintf(
+			"%s/cpu%d/topology/physical_package_id",
```
Comment (fromanirh) [inline pkg/cpu/cpu_linux.go:102]: could you please add a comment to clarify why this is the right file to learn about the cpu id in the sysfs topology? https://www.kernel.org/doc/html/v5.5/admin-guide/cputopology.html

File: pkg/cpu/cpu_linux.go [hunk 6]
```diff
@@ -64,85 +84,69 @@ func processorsGet(ctx *context.Context) []*Processor {
...
-				cores = append(cores, core)
+			if len(procAttrs[lcoreID]["vendor_id"]) != 0 {
+				proc.Vendor = procAttrs[lcoreID]["vendor_id"]
+			} else if len(procAttrs[lcoreID]["isa"]) != 0 { // RISCV64
+				proc.Vendor = procAttrs[lcoreID]["isa"]
 			}
+			procs = append(procs, proc)
+		}
+
+		// Fetch Core ID
```
Comment (fromanirh) [inline pkg/cpu/cpu_linux.go:133]: same comment about building paths

**Review by fromanirh — COMMENTED** [review]
initial review
first of all, thanks for working in this area. I think this is a welcome change.
To move forward, there are a couple of things to address:
1. we would need some aarch64 and riscv64 machines to run at least some CI on; unfortunately, code outside CI bitrots pretty fast. I'm not aware of option (not sure gh actions run on arm, IIRC not)
2. it would be very useful to have some docs highligthing how the topology exposed by sysfs exports data in case of x86_64 and risc64 and others. E.g., what's a "core" for riscv64? is it the same thing we call "core" in x86_64?

**Comment by Itxaka** [comment]
FYI you can run tests on qemu arm64 under github ci. Yes, its slow as hell, and you need to ssh into the VM to run your tests, but seems to work as expected, we run our battery of tests on it, even installing and restarting the vm to test different things and we can say it works.

See a vagrantfile to bring machine X up on qemu-system-aarch64: https://github.com/rancher-sandbox/cOS-toolkit/blob/master/tests/Vagrantfile#L30 and the makefile that brings it up: https://github.com/rancher-sandbox/cOS-toolkit/blob/master/make/Makefile.test#L61

Let me know if you need any help in this area (testing on aarch64 vms on github ci!) :)

**Comment by jak3kaj** [comment]
This PR resolves CPU discovery issues on an ARMv7 network switch as well.

I would like to see this PR merged into main.

**Comment by fromanirh** [comment]
adding arm support would be very nice indeed. Unfortunately we still need to sort out the CI and the supportability aspects, because we very much want to avoid the feature to break (or to break silently!) when the projects moves on.

**Comment by jak3kaj** [comment]
Understood.  

I'm going to work on adding arm support for some of the data that's usually stored in dmi on x86.  I will also work on setting up some arm CI in a fork and I'll circle back to see if that will be suitable for this project too.

**Comment by glimchb** [comment]
FYI @jaypipes merged #333 , this can be closed

**Comment by Itxaka** [comment]
nice!
